### DISCUSSION TITLE
Accessing config or metadata from OutputContext

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#17811] [Q&A]
hi all, i have a dagster sensor set up to detect new files in an ADLS container, then execute a job for each new file. the job involves moving datasets through multiple folders throughout the transforms process, so i’ve set up a function that gets the destination folder path based on the upstream asset’s filepath. This function lives in my IO manager so it is important that my IO manager is able to access the origin filepath from the sensor’s run_key/config

my problem is that i can’t seem to access any config or metadata from the OutputContext.
1. I first set the filepath as the run_key to the sensor, but i’m unable to access this in the OutputContext. 
2. Then i tried to use `context.add_output_metadata` from within the asset to add the tag `dagster/run_key` as metadata, but logging the `output_context.metadata` gives me an empty dict still
3. Lastly I tried to add a `RunConfig` to the sensor, but logging `output_context.config` gives me an empty dict
would appreciate any help, thanks in advance!!

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1699446212248019?thread_ts=1699446212.248019&cid=C01U954MEER).*

**Comment by hdotking** [comment]
Same problem here!

When i create an asset the metadata is contained in the _metadata key’s value as shown below:
```
#This yielded from my Dagster @asset() 
output = Output(emails_batch, metadata={"batch_index": str(batch_index)})
print(vars(output))
>>
vars(output)={...'_metadata': {'batch_index': TextMetadataValue(text='0')}}
```

But as soon as it has been yielded to the IOManager, the  `_metadata` becomes an empty `dict`.
```
vars(context)={'_step_key': 'get_json_chunks', '_name': 'result', '_job_name': '__ASSET_JOB_0', '_run_id': 'abc33433-247b-4ca1-957c-7238b6f30d43', '_metadata': {}
```

What happened to my metadata?

**Comment by jamiedemaria** [comment]
> I first set the filepath as the run_key to the sensor, but i’m unable to access this in the OutputContext. 

This makes sense/is intended. The `run_key` is just used as a mechanism to not execute two runs with the same `run_key` (ie not do duplicative work)

> Then i tried to use context.add_output_metadata from within the asset to add the tag dagster/run_key as metadata, but logging the output_context.metadata gives me an empty dict still

This one is a bit tricky, and I'm not totally sure why the output metadata isn't available in the I/O manager. That might be a feature we need to add. 

> Lastly I tried to add a RunConfig to the sensor, but logging output_context.config gives me an empty dict

This is because `output_context.config` would return the configuration for the I/O manager, not for the output being processed by the I/O manager. The `output_context` contains a `step_context` which is the context that was available on the `asset`. If you pass the path using run config, you can access the value in the I/O manager using `output_context.step_context.op_config`

**Comment by hdotking** [comment]
Sorry,  I don't quite understand. Why does yielding a `result` with metadata via the Dagster `Output()` method NOT pass that very metadata to a linked IoManager? Please let me know if i've made a mistake or misconfigured it somehow.

# Sample Code
```
# Asset definition that uses the IOManager to write chunks of emails
@asset(
    group_name="api_assets",
    partitions_def=users_partitions_def,
    io_manager_key="json_chunk_io_manager",
    description="Get json data from gmail API and save locally",
)
def get_json_chunks(context):
    batch_index = 1
    a_batch = get_batch()
    # Here we pass the batch_index as metadata to be used by the IOManager
    result = Output(a_batch, metadata={"batch_index": str(batch_index)})
    logger.info(f"Yielding output {vars(result)}")
    yield result
```

```
#IoManager

class JsonChunkIOManagerConfig(Config):
    base_dir: str


class JsonChunkIOManager(IOManager):
    def __init__(self, config: JsonChunkIOManagerConfig):
        self.base_dir = config.base_dir
        self.configs = {user: create_config(user) for user in MoonfireUser}

    def handle_output(self, context: OutputContext, obj):
        print(f"{context=}")
        print.info(f"{vars(context)=}")
        save_data()
        
    def load_input(self, context: InputContext):
        pass

@io_manager(config_schema={"base_dir": str})
def json_chunk_io_manager(context):
    # Create an instance of JsonChunkIOManagerConfig with the provided base_dir
    json_chunk_manager_config = JsonChunkIOManagerConfig(base_dir=context.resource_config["base_dir"])
    # Pass the config instance when creating JsonChunkIOManager
    return JsonChunkIOManager(config=json_chunk_manager_config)
```

```
# run_materialize.py
from dagster import materialize

from directories import get_repo_root_dir
from assets.import get_json_chunks

if __name__ == "__main__":
    result = materialize(
        [get_json_chunks],
        run_config={
            "resources": {
                "json_chunk_io_manager": {
                    "config": {
                        "base_dir": get_repo_root_dir(),
                    }
                }
            }
        },
    )
```

**Reply by jamiedemaria** [reply]
I'm not quite sure either - it might be a bug. I'll need to investigate

**Reply by hdotking** [reply]
Thanks @jamiedemaria. Let me know if I can help you solve this in any way. It's a bit of a blocker for me at the moment.

**Reply by hdotking** [reply]
@jamiedemaria any updates?

**Reply by jamiedemaria** [reply]
I've got a couple other large projects I'm working on at the moment, but this is next on my list

**Reply by yuhan** [reply]
seems relevant to https://github.com/dagster-io/dagster/issues/9076?
